### ISSUE TITLE
NPM crashes when local module is not in a subdirectory  (with minimal example)

### ISSUE THREAD

**Issue by laino** [npm/npm#18746]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

NPM crashes when a local module is not in a subdirectory of the parent module including it.

```
> uglifyjs-webpack-plugin@0.4.6 postinstall /home/lain/git/npm-bug/other/node_modules/uglifyjs-webpack-plugin
> node lib/post_install.js

npm ERR! /home/lain/git/npm-bug/other/node_modules/fsevents/node_modules/abbrev is not a child of /home/lain/git/npm-bug/main
```

I've seen it crash in [various other funny ways](https://github.com/npm/npm/issues/18403#issuecomment-334958575) depending on various factors, but hopefully they all have the same root cause.

#### Root Cause

The root cause appears to be that NPM erroneously creates a ```node_modules``` folder in the directory of the local dependency. NPM doesn't do this when the dependency is in a subdirectory of its parent, for example ``file:./other`` instead of ``file:../other``.

#### How can the CLI team reproduce the problem?

Clone this repository and have a look at ``README.md`` and ``reproduce.sh`` script:

https://github.com/laino/npm-bug

### supporting information:

 - `npm -v` prints: ```5.4.2``` (also confirmed to happen on 5.3.0)
 - `node -v` prints: ```v8.6.0```
 - `npm config get registry` prints: ```https://registry.npmjs.org/```
 - Windows, OS X/macOS, or Linux?: Linux

**Comment by kriswill** [comment]
The demo perfectly reproduces the bug.  This arrangement of sibling directories is very common in node.js monorepos.

**Comment by chrismwendt** [comment]
**tl;dr** This has been fixed in npm 5.6.0 https://github.com/npm/npm/releases/tag/v5.6.0 in https://github.com/npm/npm/pull/19054

I did some digging into this and found that the problem occurs when a dependency points to a package outside the current directory (`"other": "file:../other"`) and that package depends on a package that will not be installed on the current OS (`other` -> `webpack` -> `watchpack` -> `chokidar` -> `fsevents`, where `fsevents` is only installed on Mac).

The [`rollbackFailedOptional`](https://github.com/npm/npm/blob/d3095ff20b8ea01e7fbf93a4a697a04fea77d8e6/lib/install.js#L574) cleanup routine in npm uses `fs-vacuum` to remove the optional dependency, but that confuses me because npm shouldn't have installed `fsevents` in the first place.

**Comment by laino** [comment]
I can confirm it's fixed 5.6.0

Closing this.
